### ISSUE TITLE
[OSS] Handle torch AMP auto skip inf/nan steps

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/fairscale#180]
## 🚀 Feature
Using torch amp can lead to step() being skipped in the presence of inf or nans. An issue with that with OSS is that each rank will see a different step, meaning that some ranks could skip and others try to step (typically vanishing gradient on the way up and underflow with fp16). We need to handle that properly

## Motivation
Without that using torch amp with OSS can easily deadlock, see for instance https://github.com/facebookresearch/fairscale/pull/179

## Pitch
A somewhat elegant and not too far reaching solution to that could be to add a config flag to OSS, following which each rank will check for inf/nan on its grads, and signal all ranks that they should skip step() in that case. Users willing to apply OSS and torch amp would be advised to switch this flag on to prevent deadlocks

## Alternatives
Prevent mixed precision and OSS, but that makes OSS a lot less compelling

## Additional context
#179 and https://pytorch.org/docs/stable/notes/amp_examples.html?highlight=mixed%20precision#id2

**Comment by blefaudeux** [comment]
hmm, I'm wondering now if subclassing Scaler would not be a better idea, so that we check on all grads before hypothetically skipping the step()

**Comment by SeanNaren** [comment]
Hey @blefaudeux I think we're in sync on this, I'm currently looking at subclassing the scaler and ensuring the inf check is reduced correctly:

```python
import torch.distributed as dist
from torch.cuda.amp import GradScaler


class ShardedGradScaler(GradScaler):
    def step(self, optimizer, *args, **kwargs):
        optimizer_state = self._per_optimizer_states[id(optimizer)]

        for v in optimizer_state["found_inf_per_device"].values():
            dist.all_reduce(v)

        return super().step(optimizer, *args, **kwargs)
```

**Comment by blefaudeux** [comment]
Agreed, so if OSS can expose a distributed gradient check (not too different from the clip_grad_norm) the ShardedGradScaler code would be fairly minimal and easy to grok

**Comment by blefaudeux** [comment]
looks like there are some subtleties to keep in mind for speed (for instance prevent the upstream GradScaler from spending too long checking for infs again in the `super().step()`), but doable. I should have a working example along these lines in the next couple of hours

**Comment by blefaudeux** [comment]
> Hey @blefaudeux I think we're in sync on this, I'm currently looking at subclassing the scaler and ensuring the inf check is reduced correctly:
> 
> ```python
> import torch.distributed as dist
> from torch.cuda.amp import GradScaler
> 
> 
> class ShardedGradScaler(GradScaler):
>     def step(self, optimizer, *args, **kwargs):
>         optimizer_state = self._per_optimizer_states[id(optimizer)]
> 
>         for v in optimizer_state["found_inf_per_device"].values():
>             dist.all_reduce(v)
> 
>         return super().step(optimizer, *args, **kwargs)
> ```

looks like this take is the cleanest, I just spend some time getting up to speed with GradScaler and it looks like the most idiomatic way indeed.
